typescript/sdk/src/providers/MultiProvider.ts

```diff
@@ -21,38 +21,58 @@ import { pick } from '../utils/objects';
 type Provider = providers.Provider;
 
 export class MultiProvider {
+  private readonly logger: Debugger = debug('hyperlane:MultiProvider');
   public readonly metadata: ChainMap<ChainMetadata> = {};
   private readonly providers: ChainMap<Provider> = {};
-  private readonly signers: ChainMap<Signer> = {};
-  private readonly logger: Debugger = debug('hyperlane:MultiProvider');
-  private sharedSigner: Signer | null = null; // A single signer to be used for all chains
+  private signers: ChainMap<Signer> = {};
+  private useSharedSigner = false; // A single signer to be used for all chains
 
   /**
    * Create a new MultiProvider with the given chainMetadata,
    * or the SDK's default metadata if not provided
    */
   constructor(chainMetadata: ChainMap<ChainMetadata> = defaultChainMetadata) {
     this.metadata = chainMetadata;
+    // Ensure no two chains have overlapping names/domainIds/chainIds
+    const chainNames = new Set<string>();
+    const chainIds = new Set<number>();
+    const domainIds = new Set<number>();
+    for (const chain of Object.values(chainMetadata)) {
+      const { name, chainId, domainId } = chain;
+      if (chainNames.has(name))
+        throw new Error(`Duplicate chain name: ${name}`);
+      if (chainIds.has(chainId))
+        throw new Error(`Duplicate chain id: ${chainId}`);
+      if (domainIds.has(chainId))
+        throw new Error(`Overlapping chain/domain id: ${chainId}`);
+      if (domainId && domainIds.has(domainId))
+        throw new Error(`Duplicate domain id: ${domainId}`);
+      if (domainId && chainIds.has(domainId))
+        throw new Error(`Overlapping chain/domain id: ${domainId}`);
+      chainNames.add(name);
+      chainIds.add(chainId);
+      if (domainId) domainIds.add(domainId);
+    }
   }
 
   /**
-   * Get the metadata for a given chain name or chain id
+   * Get the metadata for a given chain name, chain id, or domain id
    * @throws if chain's metadata has not been set
    */
   tryGetChainMetadata(chainNameOrId: ChainName | number): ChainMetadata | null {
     let chainMetadata: ChainMetadata | undefined;
     if (typeof chainNameOrId === 'string') {
       chainMetadata = this.metadata[chainNameOrId];
```

Comment:
Why not return directly, why assign to `chainMetadata` first?

Reply:
Because I want to ensure I return `null` instead of `undefined` without repeating `|| null` in 2 places